rfd/0144-client-tools-updates.md

```diff
@@ -0,0 +1,309 @@
+---
+authors: Russell Jones (rjones@goteleport.com) and Bernard Kim (bernard@goteleport.com)
+state: draft
+---
+
+# RFD 0144 - Client Tools Updates
+
+## Required Approvers
+
+* Engineering: @sclevine && @bernardjkim && @r0mant
+* Product: @klizhentas || @xinding33
+* Security: @reedloden
+
+## What/Why
+
+This RFD describes how client tools (like `tsh` and `tctl`) can be kept up to
+date, either using automatic updates or self-managed updates.
+
+Keeping client tools updated helps with security (fixes for known security
+vulnerabilities are pushed to endpoints), bugs (fixes for resolved issues are
+pushed to endpoints), and compatibility (users no longer have to learn and
+understand [Teleport component
+compatibility](https://goteleport.com/docs/upgrading/overview/#component-compatibility)
+rules).
+
+## Details
+
+### Summary
+
+Client tools (like `tsh` and `tctl`) will automatically download and install
+the version of client tools recommended by the Teleport cluster.
+
+The `cluster_maintenance_config` resource will be updated to include a
+`tools_version` field to control the version of client tools advertised by the
+cluster. This resource will not be configurable on Teleport Cloud. It will be
+managed by the Teleport Cloud team.
+
+If cluster administrators want to opt-out of automatic updates and manage
+updates themselves, they will be able to watch the `cluster_maintenance_config`
+resource for changes and push out updates to endpoints manually.
+
+Inspiration drawn from https://go.dev/doc/toolchain.
+
+### Implementation
+
+#### Client tools
+
+##### Automatic updates
+
+The first time a user executes `tsh login` with a version of `tsh` that
+supports automatic updates, the user will be prompted to enroll in automatic
+updates. This choice will be saved in client tools configuration.
+
+```
+$ tsh --proxy=proxy.example.com login
+Keep client tools like tsh and tctl automatically updated? [YES/no]
+
+[...]
+```
+
+```
+$ cat ~/.tsh/config/config.yaml
+enable_autoupdate: true | false
+```
+
+Once enrolled, upon login client tools will check the `tools_version`
+advertised by the cluster at `/v1/webapi/ping`. If this version differs from
+that of the running binary, the recommended version will be downloaded and
+installed.
+
+The original binaries will not be overwritten by automatic updates, instead
+per-cluster binaries with permissions `0555` will be stored at
+`~/.tsh/bin/proxyName/{tctl,tsh}`. This will act as both a cache and allow
+users to connect to different Teleport clusters (running different versions of
+Teleport) without juggling multiple versions of client tools.
```

Comment:
We might want to have a lock so two `tsh`/`tctl` commands don't attempt to download the binary simultaneously, nor attempt to exec an incomplete binary. This often happens when shims like this are used in scripts, e.g. in CI.

Reply:
Yep, had that in mind, forgot to add it to the RFD. Will do.